### PR TITLE
fv3fit.emulation: add temperature scaling

### PR DESCRIPTION
**Pull request by nbren12**
The microphysics emulator performs poorly in cold places. Performing ML in a temperature-scaled space should help.

Added public API:
- `fv3fit.emulation.transforms.ConditionallyScaled` configuration class

Some interpolation and groupby-apply-combine functions utilties:
- `fv3fit.keras.math.piecewise`
- `fv3fit.keras.math.groupby_bins`

### CODE REVIEW

**Comment by nbren12** [comment]
Running an experiment with this code here: https://wandb.ai/ai2cm/microphysics-emulation/runs/1ctz7e3w/overview?workspace=user-nbrenowitz

File: external/fv3fit/fv3fit/emulation/transforms/factories.py [hunk 1]
```diff
@@ -45,6 +49,83 @@ def build(self, sample: TensorDict) -> TensorTransform:
...
+    values = groupby_bins(edges, x, y, reduction)
+
+    def interp(x: tf.Tensor) -> tf.Tensor:
+        return piecewise(edges[:-1], values, x)
+
+    return interp
+
+
+@dataclasses.dataclass
+class ConditionallyScaled(TransformFactory):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/transforms/factories.py:72]: Can you also indicate in the class name that this is for a residual?

File: external/fv3fit/fv3fit/emulation/transforms/transforms.py [hunk 2]
```diff
@@ -53,6 +53,45 @@ def backward(self, y: TensorDict) -> TensorDict:
         return out
 
 
+class ConditionallyScaledTransform(TensorTransform):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/transforms/transforms.py:56]: Similar to the new factory class, can you make it clear in the name that this is for residuals?

File: external/fv3fit/fv3fit/keras/math.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        xg: points to interpolate at
+
+    Return:
+        yg: interpolated values. same shape as xg. Uses constant extrapolation.
+            Within the min/max of x it has the following form::
+
+                f(xg) = y[i] if x[i] <= xg < x[i+1]
+    """
+    xg_flat = tf.reshape(xg, [-1])
+    index = tf.searchsorted(x, xg_flat, side="right") - 1
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/math.py:20]: Will this raise an error if `x` is not actually sorted, or just silently return the wrong output?

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 4]
```diff
@@ -89,7 +90,9 @@ class TrainConfig:
     test_url: str
     out_url: str
     transform: TransformConfig = field(default_factory=TransformConfig)
-    tensor_transform: List[TransformedVariableConfig] = field(default_factory=list)
+    tensor_transform: List[
+        Union[TransformedVariableConfig, ConditionallyScaled]
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/train_microphysics.py:94]: Maybe `Union[TransformedVariableConfig, ConditionallyScaled]` can be defined outside of this class as `Transforms` or something.

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 5]
```diff
@@ -234,7 +237,13 @@ def open_dataset(
         self, url: str, nfiles: Optional[int], required_variables: Set[str],
     ) -> tf.data.Dataset:
         nc_open_fn = self.transform.get_pipeline(required_variables)
-        return nc_dir_to_tf_dataset(url, nc_open_fn, nfiles=nfiles)
+        return nc_dir_to_tf_dataset(
+            url,
+            nc_open_fn,
+            nfiles=nfiles,
+            shuffle=True,
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/train_microphysics.py:244]: Why shuffle?

File: external/fv3fit/fv3fit/emulation/transforms/factories.py [hunk 6]
```diff
@@ -45,6 +49,83 @@ def build(self, sample: TensorDict) -> TensorTransform:
...
+        min_scale: the minimium scale to normalize by. Used when the scale might
+            be 0.
+
+    """
+
+    to: str
+    condition_on: str
+    bins: int
+    field: Field
+    min_scale: float = 0.0
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/transforms/factories.py:97]: Would it be easier to just make the default some epsilon in the case of the scale being zero?

**Review by AnnaKwa — APPROVED** [review]
LGTM! Just some minor comments/questions.

File: external/fv3fit/fv3fit/keras/math.py (hunk 3 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/keras/math.py:20]: It may silently fail, but I decided to let tensorflow make it's own complaints if it wants to. I did put an assertion in the `groupby` to ensure we don't do that in practice.

File: external/fv3fit/fv3fit/train_microphysics.py (hunk 4 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/train_microphysics.py:94]: If they start getting used elsewhere sure, but this Union is only used for dacite really. I used the `TransformFactory` protocol  for type-checking most places.

File: external/fv3fit/fv3fit/train_microphysics.py (hunk 5 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/train_microphysics.py:244]: We weren't shuffling before, so only Jan data was used to fit all the normalization. It's possible this could have an impact since it is southern summer, and the coldest temperatures are not present.

File: external/fv3fit/fv3fit/emulation/transforms/factories.py (hunk 6 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/transforms/factories.py:97]: It would be nice if the default worked for all variables---including ones with smaller non-zero magnitude than `min_scale`. I think the most sane default is to do nothing at all.

File: external/fv3fit/fv3fit/emulation/transforms/transforms.py [hunk 7]
```diff
@@ -53,6 +53,45 @@ def backward(self, y: TensorDict) -> TensorDict:
...
+        self,
+        to: str,
+        input_name: str,
+        output_name: str,
+        on: str,
+        scale: Callable[[tf.Tensor], tf.Tensor],
+        center: Callable[[tf.Tensor], tf.Tensor],
+        min_scale: float = 0.0,
+    ) -> None:
+        self.to = to
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/transforms/transforms.py:67]: Whoops. I failed to push the commit before merging...maybe we can just let this one slide? (or I could open another PR). Let me know.

[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/emulation/transforms/transforms.py:67]: That's fine, I think it's usefulness is mostly for reviewers looking at this class. If you make a future PR that touches this class you can just add it then.
